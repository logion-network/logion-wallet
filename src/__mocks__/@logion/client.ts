import { CheckResultType } from "@logion/client/dist/CollectionItem.js";
import { DefaultSignAndSendStrategy } from "@logion/client/dist/Signer.js";

import {
    LogionClient,
    AccountTokens,
    toIsoString,
    fromIsoString,
    ClosedCollectionLoc,
    DraftRequest,
    EditableRequest,
    OpenLoc,
    LocRequestState,
    LocsState,
    HashOrContent,
} from '../LogionClientMock';

export {
    LogionClient,
    AccountTokens,
    toIsoString,
    fromIsoString,
    CheckResultType,
    ClosedCollectionLoc,
    DraftRequest,
    EditableRequest,
    OpenLoc,
    DefaultSignAndSendStrategy,
    LocRequestState,
    LocsState,
    HashOrContent,
}

export { MimeType } from "@logion/client/dist/Mime.js";
