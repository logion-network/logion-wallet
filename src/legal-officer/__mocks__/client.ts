import {
    closeLoc,
    deleteLink,
    voidLoc,
    getVerifiedThirdPartySelections,
} from "./ClientMock";

export {
    closeLoc,
    deleteLink,
    voidLoc,
    getVerifiedThirdPartySelections,
}
