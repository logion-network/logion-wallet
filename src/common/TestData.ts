import { LegalOfficer, PostalAddress, UserIdentity } from '@logion/client';
import { Transaction } from '@logion/client/dist/TransactionClient.js';
import { Coin, CoinBalance, SYMBOL } from '@logion/node-api/dist/Balances.js';
import { PrefixedNumber, ATTO } from '@logion/node-api/dist/numbers.js';

import { ColorTheme, rgbaToHex } from './ColorTheme';

export const DEFAULT_LEGAL_OFFICER = "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY"; // Alice
export const ANOTHER_LEGAL_OFFICER = "5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty"; // Bob
export const A_THIRD_LEGAL_OFFICER = "5FLSigC9HGRKVhB9FiEo4Y3koPsNmBmLJbpXg2mp1hXcS59Y"; // Charlie

export const DEFAULT_IDENTITY: UserIdentity = {
    firstName: "John",
    lastName: "Doe",
    email: "john.doe@logion.network",
    phoneNumber: "+1234",
};

export const DEFAULT_ADDRESS: PostalAddress = {
    line1: "Place de le République Française, 10",
    line2: "boite 15",
    postalCode: "4000",
    city: "Liège",
    country: "Belgium",
};

export const COLOR_THEME: ColorTheme = {
    type: 'light',
    shadowColor: rgbaToHex('#3b6cf4', 0.1),
    dashboard: {
        background: '#eff3fe',
        foreground: '#000000',
    },
    sidebar: {
        background: '#ffffff',
        foreground: '#000000',
        activeItemBackground: '#d8e2fd',
    },
    accounts: {
        iconBackground: '#3b6cf4',
        hintColor: rgbaToHex('#000000', 0.6),
        foreground: '#000000',
        background: 'eff3fe',
        legalOfficerIcon: {
            category: 'legal-officer',
            id: 'account-shield'
        }
    },
    frame: {
        background: '#ffffff',
        foreground: '#000000',
        link: rgbaToHex('#3b6cf4', 0.15),
        altBackground: '#eff3fe',
    },
    topMenuItems: {
        iconGradient: {
            from: '#3b6cf4',
            to: '#6050dc',
        }
    },
    bottomMenuItems: {
        iconGradient: {
            from: '#7a90cb',
            to: '#3b6cf4',
        }
    },
    recoveryItems: {
        iconGradient: {
            from: '#7a90cb',
            to: '#3b6cf4',
        }
    },
    buttons: {
        secondary: {
            background: '#eff3fe',
            foreground: '#000000',
        }
    },
    select: {
        background: '#ffffff',
        foreground: '#000000',
        menuBackgroundColor: '#ffffff',
        selectedOptionBackgroundColor: '#ffffff',
    },
    table: {
        background: '#0c163d',
        foreground: '#ffffff',
        header: {
            background: '#0c163d',
            foreground: '#ffffff',
        },
        row: {
            background: '#152665',
            foreground: '#ffffff',
        }
    },
    tabs: {
        background: '#0c163d',
        foreground: '#ffffff',
        inactiveTab: {
            background: '#d8e2fd',
            foreground: '#ffffff',
        },
        borderColor: '#3b6cf4',
    },
    dialog: {
        background: '#eff3fe',
        foreground: '#000000',
        borderColor: '#e6007a',
    },
    dialogTable: {
        background: '#0c163d',
        foreground: '#ffffff',
        header: {
            background: '#0c163d',
            foreground: '#ffffff',
        },
        row: {
            background: '#152665',
            foreground: '#ffffff',
        }
    },
};

export const DEFAULT_COIN: Coin = {
    id: 'lgnt',
    iconId: 'lgnt',
    iconType: 'svg',
    name: "Logion",
    symbol: SYMBOL
};

export const DEFAULT_COIN_BALANCE: CoinBalance = {
    coin: DEFAULT_COIN,
    balance: new PrefixedNumber("42", ATTO),
    available: new PrefixedNumber("42", ATTO),
    level: 0.1,
};

export const DEFAULT_TRANSACTION: Transaction = {
    id: "id",
    from: "from",
    to: "to",
    pallet: "pallet",
    method: "method",
    transferValue: "420",
    tip: "0",
    fee: "125000",
    reserved: "0",
    total: "125420",
    createdOn: "2021-07-28T12:30:00.000",
    type: "Sent",
    transferDirection: "Sent",
    successful: true
}

export const DEFAULT_FAILED_TRANSACTION:Transaction = {
    ...DEFAULT_TRANSACTION,
    successful: false,
    error: {
        section: 'aSection',
        name: 'aName',
        details: 'someDetails'
    }
}

export const PATRICK: LegalOfficer = {
    name: "Patrick Gielen",
    address: DEFAULT_LEGAL_OFFICER,
    additionalDetails: "",
    userIdentity: {
        firstName: "Patrick",
        lastName: "Gielen",
        email: "patrick@logion.network",
        phoneNumber: "+32 498 237 107",
    },
    postalAddress: {
        company: "MODERO",
        line1: "Huissier de Justice Etterbeek",
        line2: "Rue Beckers 17",
        postalCode: "1040",
        city: "Etterbeek",
        country: "Belgique"
    },
    node: "http://logion.patrick.com",
    logoUrl: "",
    nodeId: "",
};

export const GUILLAUME: LegalOfficer = {
    name: "Guillaume Grain",
    address: ANOTHER_LEGAL_OFFICER,
    additionalDetails: "",
    userIdentity: {
        firstName: "Patrick",
        lastName: "Gielen",
        email: "g.grain@adrastee-lyon.fr",
        phoneNumber: "+33 4 78 52 87 56",
    },
    postalAddress: {
        company: "SELARL ADRASTEE",
        line1: "Gare des Brotteaux",
        line2: "14, place Jules Ferry",
        postalCode: "69006",
        city: "LYON",
        country: "France"
    },
    node: "http://logion.guillaume.com",
    logoUrl: "",
    nodeId: "",
};

export const ALAIN: LegalOfficer = {
    name: "Alain Barland",
    address: A_THIRD_LEGAL_OFFICER,
    additionalDetails: "",
    userIdentity: {
        firstName: "Patrick",
        lastName: "Gielen",
        email: "alain.barland@auxilia-conseils.com",
        phoneNumber: "+33 2 48 67 50 50",
    },
    postalAddress: {
        company: "AUXILIA CONSEILS 18",
        line1: "Huissiers de Justice associés",
        line2: "7 rue Jean Francois Champollion Parc Comitec",
        postalCode: "18000",
        city: "Bourges",
        country: "France"
    },
    node: "http://logion.alain.com",
    logoUrl: "",
    nodeId: "",
};

export const legalOfficers: LegalOfficer[] = [
    PATRICK,
    GUILLAUME,
    ALAIN
];
