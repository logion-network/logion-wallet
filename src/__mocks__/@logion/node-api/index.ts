import { ApiPromise } from "src/__mocks__/PolkadotApiMock";
export * from "@logion/node-api/dist/UUID";

export function buildApi() {
    return new ApiPromise();
}

export * from "./dist/LogionLoc";
