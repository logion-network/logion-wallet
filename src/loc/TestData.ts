import { LegalOfficerCase, UUID, Hash } from "@logion/node-api";
import { LocData, HashString } from "@logion/client";

export function buildLocRequest(locId: UUID, loc: LegalOfficerCase): LocData {
    return {
        ownerAddress: loc.owner,
        requesterAddress: loc.requesterAddress,
        requesterLocId: loc.requesterLocId,
        description: "Description",
        locType: loc.locType,
        createdOn: "",
        decisionOn: "",
        id: locId,
        status: loc.closed ? "CLOSED" : "OPEN",
        files: loc.files.map((locFile, index) => ({
            ...locFile,
            nature: HashString.fromValue(`File ${index}`),
            name: `File ${index}`,
            addedOn: "",
            published: false,
            restrictedDelivery: false,
            contentType: "text/plain",
            size: 42n,
            status: "DRAFT",
        })),
        metadata: loc.metadata.map((locFile, index) => ({
            ...locFile,
            name: HashString.fromValue(`Data ${index}`),
            value: HashString.fromValue(`Value ${index}`),
            addedOn: "",
            published: false,
            status: "DRAFT",
        })),
        links: loc.links.map((locFile, index) => ({
            ...locFile,
            name: `Link ${index}`,
            nature: HashString.fromValue(`Nature ${index}`),
            addedOn: "",
            target: locFile.id.toString(),
            published: false,
            status: "DRAFT",
        })),
        closed: false,
        verifiedIssuer: false,
        issuers: [],
    };
}
