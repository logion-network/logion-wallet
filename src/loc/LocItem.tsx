import {
    LogionClient,
    LocRequestState,
    LocData,
    LocsState,
    EditableRequest,
    MergedLink,
    PublicLoc,
    ItemStatus,
    ReviewResult,
    HashString,
    OpenLoc,
    PendingRequest,
} from "@logion/client";
import { UUID, LocType, Fees, ValidAccountId, Hash } from "@logion/node-api";
import { ReactNode, useCallback } from "react";
import { CallCallback } from "src/ClientExtrinsicSubmitter";
import { Viewer } from "src/common/CommonContext";
import { isGrantedAccess } from "src/common/Model";
import { Rules } from "src/common/Responsive";
import SubmitterName from "src/common/SubmitterName";
import { Cell, Column, DateTimeCell } from "src/common/Table";
import { Child } from "src/common/types/Helpers";
import ViewFileButton from "src/common/ViewFileButton";
import { documentClaimHistoryPath } from "src/legal-officer/LegalOfficerPaths";
import { fullCertificateUrl } from "src/PublicPaths";
import { documentClaimHistoryPath as userDocumentClaimHistoryPath } from "src/wallet-user/UserRouter";
import LocLinkDetails from "./LocLinkDetails";
import LocPrivateFileDetails from "./LocPrivateFileDetails";
import LocPublicDataDetails from "./LocPublicDataDetails";
import RestrictedDeliveryCell from "./RestricedDeliveryCell";
import { ContributionMode } from "./types";
import ReviewStatusCell from "./ReviewStatusCell";
import HelpTooltip from "src/components/helptooltip/HelpTooltip";
import { DateTime } from "luxon";

export type LocItemType = 'Data' | 'Document' | 'Linked LOC';

export interface LinkData {
    readonly linkedLoc: LocData;
    readonly linkDetailsPath: string;
    readonly nature: HashString;
}

export interface MetadataData {
    readonly name: HashString;
    readonly value: HashString;
}

export interface FileData {
    readonly hash: Hash;
    readonly fileName: string;
    readonly nature: HashString;
    readonly size: bigint;
    readonly storageFeePaidBy: string;
}

export interface CommonData {
    readonly timestamp: DateTime | null;
    readonly type: LocItemType;
    readonly submitter?: ValidAccountId;
    readonly status: ItemStatus;
    readonly newItem: boolean;
    readonly template: boolean;
    readonly isSet?: boolean;
    readonly fees?: Fees;
    readonly reviewedOn?: DateTime;
    readonly rejectReason?: string;
    readonly acknowledgedByOwner?: boolean;
    readonly acknowledgedByVerifiedIssuer?: boolean;
    readonly defaultTitle?: string;
}

abstract class AbstractLocItem<T> implements CommonData {

    constructor(commonData: CommonData, data?: T) {
        this.commonData = commonData;
        this._data = data;
    }

    protected commonData: CommonData;
    private readonly _data?: T;

    get timestamp() {
        return this.commonData.timestamp;
    }

    get type() {
        return this.commonData.type;
    }

    get submitter() {
        return this.commonData.submitter;
    }

    get status() {
        return this.commonData.status;
    }

    get newItem() {
        return this.commonData.newItem;
    }

    get template() {
        return this.commonData.template;
    }

    get isSet() {
        return this.commonData.isSet;
    }

    get fees() {
        return this.commonData.fees;
    }

    get reviewedOn() {
        return this.commonData.reviewedOn;
    }

    get rejectReason() {
        return this.commonData.rejectReason;
    }

    get acknowledgedByOwner() {
        return this.commonData.acknowledgedByOwner;
    }

    get acknowledgedByVerifiedIssuer() {
        return this.commonData.acknowledgedByVerifiedIssuer;
    }

    get defaultTitle(): string  {
        return this.commonData.defaultTitle || this.commonData.type;
    }

    isPublishedOrAcknowledged(): boolean {
        return this.commonData.status === "PUBLISHED" || this.commonData.status === "ACKNOWLEDGED"
    }

    abstract title(): string;

    hasData(): boolean {
        return this.commonData.submitter !== undefined && this._data !== undefined;
    }

    data(): T {
        if(this._data) {
            return this._data;
        } else {
            throw new Error("Data not set");
        }
    }

    as<U>(): U {
        return this._data as U;
    }

    abstract publish(timestamp: DateTime | null, fees?: Fees, storageFeePaidBy?: string): AbstractLocItem<T>;

    get submitterOrThrow() {
        if(this.commonData.submitter) {
            return this.commonData.submitter;
        } else {
            throw new Error("Data not set");
        }
    }
}

export class MetadataItem extends AbstractLocItem<MetadataData> {

    static TYPE: LocItemType = "Data";

    override title() {
        if (this.hasData()) {
            return this.data().name.value || "-";
        } else {
            return this.defaultTitle;
        }
    }

    override publish(timestamp: DateTime | null, fees?: Fees, _storageFeePaidBy?: string): MetadataItem {
        return new MetadataItem(
            {
                ...this.commonData,
                timestamp,
                fees,
            },
            this.data(),
        );
    }
}

export class FileItem extends AbstractLocItem<FileData> {

    override title() {
        if (this.hasData()) {
            return this.data().nature.value || "-";
        } else {
            return this.defaultTitle
        }
    }

    override publish(timestamp: DateTime | null, fees?: Fees, storageFeePaidBy?: string): FileItem {
        const newFileData = this.data() && storageFeePaidBy ? {
            ...this.data(),
            storageFeePaidBy,
        } : this.data();
        return new FileItem(
            {
                ...this.commonData,
                timestamp,
                fees,
            },
            newFileData,
        );
    }
}

export class LinkItem extends AbstractLocItem<LinkData> {

    override title() {
        if (this.hasData()) {
            return this.data().nature.value || "-";
        } else {
            return this.defaultTitle;
        }
    }

    override publish(timestamp: DateTime | null, fees?: Fees, _storageFeePaidBy?: string): LinkItem {
        return new LinkItem(
            {
                ...this.commonData,
                timestamp,
                fees,
            },
            this.data(),
        );
    }
}

export type LocItem = LinkItem | MetadataItem | FileItem;

export enum PublishStatus {
    NONE,
    START,
    PUBLISHING,
    DONE
}

export interface PublishState {
    status: PublishStatus;
}

export interface PublishProps {
    locItem: LocItem;
    itemType: 'Public Data' | 'Document' | 'Link';
    publishMutator: (current: LocRequestState, callback: CallCallback) => Promise<LocRequestState>;
    feesEstimator: () => Promise<Fees>;
}

export function buildItemTableColumns(args: {
    loc: LocData,
    locState: LocRequestState,
    width: (rule: Rules) => string,
    currentAddress: string | undefined,
    contributionMode: ContributionMode | undefined,
    viewer: Viewer,
    renderActions: (locItem: LocItem) => ReactNode,
}): Column<LocItem>[] {
    const {
        loc,
        locState,
        width,
        currentAddress,
        contributionMode,
        viewer,
        renderActions,
    } = args;
    let columns: Column<LocItem>[] = [
        {
            header: "Public description",
            render: locItem => <Cell content={ locItem.title() } overflowing tooltipId={`${loc.id}-name-tooltip`}/>,
            renderDetails: locItem => renderDetails(loc, locItem, viewer),
            detailsExpanded: locItem => locItem.newItem || (locItem.template && (viewer !== "LegalOfficer" || (locItem.status !== "PUBLISHED" && locItem.status !== "ACKNOWLEDGED"))),
            hideExpand: locItem => locItem.template && viewer !== "LegalOfficer",
            align: "left",
            width: width({
                onSmallScreen: '150px',
                otherwise: '250px'
            }),
        },
        {
            header: "Timestamp",
            render: locItem => <DateTimeCell
                dateTime={ locItem.timestamp }
                spinner={ locItem.status === 'PUBLISHED' } />,
            width: width({
                onSmallScreen: '100px',
                otherwise: '200px'
            }),
        },
        {
            header: "Type",
            render: locItem => <Cell content={
                <>
                    <span className="item-type">{ locItem.type }</span> {
                        locItem.type === 'Document' &&
                        canViewFile(viewer, currentAddress, locItem, contributionMode) &&
                        <ViewFileButton
                            nodeOwner={ loc.ownerAddress }
                            fileName={ locItem.as<FileData>().fileName }
                            downloader={ () => locState?.getFile(locItem.as<FileData>().hash) }
                        />
                    }
                </> } />,
            width: "160px",
            align: "left"
        },
        {
            header: "Submitted by",
            render: locItem => <SubmitterName loc={ loc } submitter={ locItem.submitter?.address } />,
        }
    ];

    if(loc.locType === "Collection" && viewer === "LegalOfficer") {
        columns.push({
            header: "Restricted Delivery?",
            render: locItem => locItem.type === "Document" && locItem.hasData() ? <RestrictedDeliveryCell hash={ locItem.as<FileData>().hash }/> : null,
            width: "130px",
        });
    }

    columns = columns.concat([
        {
            header: <>Reviewed <HelpTooltip id="reviewed" help="You confirm that the content you submitted for review by the Legal Officer is good to be recorded by the logion infrastructure with regard to the LOC context and to the specific content that will be publicly available after the recording (eg: GDPR)."/></>,
            render: locItem => <ReviewStatusCell locItem={ locItem } />,
        },
        {
            header: "",
            render: locItem => renderActions(locItem),
            width: width({
                onSmallScreen: viewer === "LegalOfficer" ? '345px' : '145px',
                otherwise: '400px'
            }),
        }
    ]);
    return columns;
}

function renderDetails(loc: LocData | undefined, locItem: LocItem, viewer: Viewer): Child {
    return (
        <>
            { locItem.type === 'Data' && <LocPublicDataDetails item={ locItem as MetadataItem } /> }
            {
                locItem.type === 'Document' &&
                <LocPrivateFileDetails
                    item={ locItem as FileItem }
                    documentClaimHistory={ loc?.locType === "Collection" && !locItem.template ? documentClaimHistory(viewer, loc, locItem.as<FileData>().hash) : undefined }
                    otherFeesPaidByRequester={ loc?.requesterLocId === undefined }
                />
            }
            { locItem.type === 'Linked LOC' && <LocLinkDetails item={ locItem as LinkItem } /> }
        </>
    )
}

function canViewFile(viewer: Viewer, address: string | undefined, item: LocItem, contributionMode?: ContributionMode): boolean {
    return item.hasData() && (contributionMode !== 'VerifiedIssuer' || item.submitterOrThrow.address === address) && (viewer === "User" || item.status !== "DRAFT");
}

export function documentClaimHistory(viewer: Viewer, loc: LocData | null, hash: Hash) {
    if(!loc) {
        return "";
    } else if(viewer === "LegalOfficer") {
        return documentClaimHistoryPath(loc.id, hash);
    } else {
        return userDocumentClaimHistoryPath(loc.id, hash);
    }
}

export function useDeleteMetadataCallback(mutateLocState: (mutator: (current: LocRequestState) => Promise<LocRequestState | LocsState>) => Promise<void>) {
    return useCallback(async (item: LocItem) => {
        await mutateLocState(async current => {
            if(current instanceof EditableRequest) {
                return current.deleteMetadata({
                    nameHash: item.as<MetadataData>().name.hash,
                });
            } else {
                return current;
            }
        });
    }, [ mutateLocState ]);
}

export function useDeleteFileCallback(mutateLocState: (mutator: (current: LocRequestState) => Promise<LocRequestState | LocsState>) => Promise<void>) {
    return useCallback(async (item: LocItem) => {
        await mutateLocState(async current => {
            if(current instanceof EditableRequest) {
                return current.deleteFile({
                    hash: item.as<FileData>().hash,
                });
            } else {
                return current;
            }
        });
    }, [ mutateLocState ]);
}

export function useDeleteLinkCallback(mutateLocState: (mutator: (current: LocRequestState) => Promise<LocRequestState | LocsState>) => Promise<void>) {
    return useCallback(async (item: LocItem) => {
        await mutateLocState(async current => {
            if(current instanceof EditableRequest) {
                return current.deleteLink({
                    target: item.as<LinkData>().linkedLoc.id,
                });
            } else {
                return current;
            }
        });
    }, [ mutateLocState ]);
}

export function canDelete(account: ValidAccountId | undefined, item: LocItem, viewer: Viewer, loc: LocData): boolean {
    return item.submitter?.address === account?.address && item.submitter?.type === account?.type
            && (loc.status === "DRAFT" || loc.status === "OPEN")
            && (item.status === "DRAFT" || item.status === "REVIEW_REJECTED");
}

export function canAdd(viewer: Viewer, loc: LocData) {
    return (viewer === "User" && (!loc.voidInfo && (loc.status === "DRAFT" || loc.status === "OPEN")))
        || (viewer === "LegalOfficer"
            && (!loc.voidInfo && loc.status === "OPEN")
            && ( loc.requesterAddress === undefined || loc.requesterAddress.type !== "Polkadot"));
}

type ActorType = "Owner" | ContributionMode;

function getActorType(viewer: Viewer, contributionMode?: ContributionMode): ActorType {
    if (viewer === "LegalOfficer") {
        return "Owner"
    } else {
        return contributionMode!
    }
}

function getSubmitterType(loc: LocData, item: LocItem): ActorType {
    const submittedByOwner = item?.submitter?.address === loc.ownerAddress && item?.submitter?.type === "Polkadot";
    const submittedByRequester = loc.requesterAddress !== undefined && item.submitter !== undefined && loc.requesterAddress.equals(item.submitter);

    if (submittedByOwner) {
        return "Owner";
    } else if (submittedByRequester) {
        return "Requester";
    } else {
        return "VerifiedIssuer";
    }
}

export function canPublish(viewer: Viewer, loc: LocData, item: LocItem, contributionMode?: ContributionMode): boolean {

    const actorType = getActorType(viewer, contributionMode);
    if (actorType === "VerifiedIssuer") {
        return false;
    }

    const submitterType = getSubmitterType(loc, item);

    const publishedByOwner = submitterType === "Owner" || loc.requesterAddress === undefined || loc.requesterAddress.type !== "Polkadot";

    const publishable = item.status === "REVIEW_ACCEPTED";

    return publishable
        && loc.status === "OPEN" && !loc.voidInfo
        && (
            (actorType === "Owner" && publishedByOwner) ||
            (actorType === "Requester" && !publishedByOwner)
        );
}

export function canAcknowledge(viewer: Viewer, loc: LocData, item: LocItem, contributionMode?: ContributionMode): boolean {

    const actorType = getActorType(viewer, contributionMode);
    if (actorType === "Requester") {
        return false;
    }

    const submitterType = getSubmitterType(loc, item);

    const acknowledgeable = item.status === "PUBLISHED";

    return acknowledgeable
        && loc.status === "OPEN" && !loc.voidInfo
        && (
            (submitterType === "VerifiedIssuer" && actorType === "VerifiedIssuer" && item.acknowledgedByVerifiedIssuer !== undefined && !item.acknowledgedByVerifiedIssuer)
            || (actorType === "Owner" && item.acknowledgedByOwner !== undefined && !item.acknowledgedByOwner)
        )
}

export function canRequestReview(viewer: Viewer, loc: LocData, item: LocItem) {
    return viewer === "User" && item.status === "DRAFT" && loc.status === "OPEN";
}

export function canReview(viewer: Viewer, loc: LocData, item: LocItem) {
    return viewer === "LegalOfficer" && item.status === "REVIEW_PENDING" && (loc.status === "REVIEW_PENDING" || loc.status === "OPEN");
}

export async function getLinkData(
    address: string | undefined,
    locsState: LocsState,
    link: MergedLink,
    detailsPath: (id: UUID, locType: LocType) => string,
    client: LogionClient,
): Promise<LinkData> {
    let linkedLoc: LocData | undefined;
    const targetId = link.target;
    let linkedLocState: LocRequestState | PublicLoc | undefined = locsState.findByIdOrUndefined(targetId);
    if (!linkedLocState) {
        linkedLocState = await client.public.findLocById({ locId: targetId });
        if(linkedLocState) {
            linkedLoc = linkedLocState.data;
        }
    } else {
        linkedLoc = linkedLocState.data();
    }

    if(!linkedLoc) {
        throw new Error("Unable to locate linked LOC");
    }

    let linkDetailsPath: string;
    if (isGrantedAccess(address, linkedLoc)) {
        linkDetailsPath = detailsPath(linkedLoc.id, linkedLoc.locType);
    } else {
        linkDetailsPath = fullCertificateUrl(linkedLoc.id);
    }
    return {
        linkedLoc,
        linkDetailsPath,
        nature: link.nature,
    };
}

export function useRequestReviewCallback(mutateLocState: (mutator: (current: LocRequestState) => Promise<LocRequestState | LocsState>) => Promise<void>) {
    return useCallback(async (locItem: LocItem) => {
        mutateLocState(async current => {
            if (current instanceof EditableRequest) {
                if (locItem.type === "Data") {
                    return current.requestMetadataReview({ nameHash: locItem.as<MetadataData>().name.hash });
                } else if (locItem.type === "Document") {
                    return current.requestFileReview({ hash: locItem.as<FileData>().hash });
                } else if (locItem.type === "Linked LOC") {
                    return current.requestLinkReview({ target: locItem.as<LinkData>().linkedLoc.id });
                } else {
                    return current;
                }
            } else {
                return current;
            }
        });
    }, [ mutateLocState ]);
}

export function useReviewCallback(mutateLocState: (mutator: (current: LocRequestState) => Promise<LocRequestState | LocsState>) => Promise<void>) {
    return useCallback(async (locItem: LocItem, decision: ReviewResult, rejectReason?: string) => {
        mutateLocState(async current => {
            if(current instanceof OpenLoc || current instanceof PendingRequest) {
                if(locItem.type === "Data") {
                    return current.legalOfficer.reviewMetadata({
                        nameHash: locItem.as<MetadataData>().name.hash,
                        decision,
                        rejectReason
                    });
                } else if(locItem.type === "Document") {
                    return current.legalOfficer.reviewFile({
                        hash: locItem.as<FileData>().hash,
                        decision,
                        rejectReason
                    });
                } else if(locItem.type === "Linked LOC") {
                    return current.legalOfficer.reviewLink({
                        target: locItem.as<LinkData>().linkedLoc.id,
                        decision,
                        rejectReason
                    });
                } else {
                    return current;
                }
            } else {
                return current;
            }
        });
    }, [ mutateLocState ]);
}
