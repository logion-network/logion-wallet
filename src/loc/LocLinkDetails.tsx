import { Col } from "../common/Grid";
import LocItemDetail from "./LocItemDetail";

import './LocItemDetails.css'
import { LinkItem } from "./LocItem";
import NewTabLink from "../common/NewTabLink";
import CopyPasteButton from "../common/CopyPasteButton";
import LocItemEstimatedFees from "./LocItemEstimatedFees";

export interface Props {
    item: LinkItem;
}

export default function LocLinkDetails(props: Props) {
    return (
        <Col className="LocItemDetails" style={ { width: "100%" } }>
            <div className="frame">
                {
                    props.item.rejectReason &&
                    <>
                        <LocItemDetail label="Rejection reason">{ props.item.rejectReason || "-" }</LocItemDetail>
                        <div className="separator"></div>
                    </>
                }
                <div className="frame-title">{ props.item.isPublishedOrAcknowledged() ? "Published data" : "Data to be published" }</div>
                <LocItemDetail label="Public Description">{ props.item.hasData() ? props.item.data().nature.validValue() : "-" }</LocItemDetail>
                <LocItemDetail label="Submitter ID" copyButtonText={ props.item.submitter?.address }>
                    { props.item.submitter?.address || "-" }
                </LocItemDetail>
                <LocItemDetail label="Linked LOC" className="linked-loc">
                    <NewTabLink href={ props.item.hasData() ? props.item.data().linkDetailsPath : undefined } iconId="loc-link">{ props.item.hasData() ? props.item.data().linkedLoc.id.toDecimalString() : undefined }</NewTabLink>
                    {
                        props.item.hasData() &&
                        <CopyPasteButton value={ props.item.data().linkedLoc.id.toDecimalString() } className="medium"/>
                    }
                </LocItemDetail>
                {
                    props.item.fees &&
                    <>
                        <div className="separator"></div>
                        <LocItemDetail label="Paid fees (LGNT)">
                            <LocItemEstimatedFees
                                fees={ props.item.fees }
                                locItem={ props.item }
                                centered={ false }
                                hideTitle={ true }
                            />
                        </LocItemDetail>
                    </>
                }
            </div>
        </Col>
    )
}
