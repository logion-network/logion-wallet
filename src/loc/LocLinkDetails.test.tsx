import { shallowRender } from "../tests";
import { TEST_WALLET_USER } from "../wallet-user/TestData";
import LocLinkDetails from "./LocLinkDetails";
import { LinkItem } from "./LocItem";
import { ItemStatus, LocData } from "@logion/client";
import { render, screen, waitFor } from '@testing-library/react';
import { UUID } from "@logion/node-api";

describe("LocLinkDetails", () => {

    function createItem(status: ItemStatus, rejectReason?: string): LinkItem {
        const linkedLocId = new UUID("20de6861-6236-4ae4-9796-5a60b5d0c43d");
        return new LinkItem(
            {
                timestamp: null,
                type: "Data",
                submitter: TEST_WALLET_USER,
                status,
                rejectReason,
                newItem: false,
                template: false,
            },
            {
                linkedLoc: {
                    id: linkedLocId
                } as LocData,
                linkDetailsPath: `https://path.to/${ linkedLocId.toString() }`,
                nature: "Some nature"
            }
        )
    }

    it("renders", () => {
        const item = createItem("DRAFT");
        const element = shallowRender(<LocLinkDetails item={ item } />);
        expect(element).toMatchSnapshot();
    });

    it("shows correct title when not published", async () => {
        const item = createItem("REVIEW_ACCEPTED");
        render(<LocLinkDetails item={ item } />);
        await waitFor(() => expect(screen.getByText("Data to be published")).toBeVisible());
    });

    it("shows correct title when published", async () => {
        const item = createItem("ACKNOWLEDGED");
        render(<LocLinkDetails item={ item } />);
        await waitFor(() => expect(screen.getByText("Published data")).toBeVisible());
    });

    it("shows reason when rejected", async () => {
        const rejectReason = "Some reject reason";
        const item = createItem("REVIEW_REJECTED", rejectReason);
        render(<LocLinkDetails item={ item } />);
        await waitFor(() => expect(screen.getByText(rejectReason)).toBeVisible());
    });
});
