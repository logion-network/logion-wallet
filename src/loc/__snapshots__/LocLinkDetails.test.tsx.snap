// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocLinkDetails renders 1`] = `
<Col
  className="LocItemDetails"
  style={
    Object {
      "width": "100%",
    }
  }
>
  <div
    className="frame"
  >
    <div
      className="frame-title"
    >
      Data to be published
    </div>
    <LocItemDetail
      label="Public Description"
    >
      Some nature
    </LocItemDetail>
    <LocItemDetail
      copyButtonText="5H4MvAsobfZ6bBCDyj5dsrWYLrA8HrRzaqa9p61UXtxMhSCY"
      label="Submitter ID"
    >
      5H4MvAsobfZ6bBCDyj5dsrWYLrA8HrRzaqa9p61UXtxMhSCY
    </LocItemDetail>
    <LocItemDetail
      className="linked-loc"
      label="Linked LOC"
    >
      <NewTabLink
        href="https://path.to/20de6861-6236-4ae4-9796-5a60b5d0c43d"
        iconId="loc-link"
      >
        43690102853837749952471838389588509757
      </NewTabLink>
      <CopyPasteButton
        className="medium"
        value="43690102853837749952471838389588509757"
      />
    </LocItemDetail>
  </div>
</Col>
`;
